src/init/AppRunner.ts

```diff
@@ -54,99 +80,81 @@ export class AppRunner {
-        (error: Error): void => {
-          // Instantiation of components has failed, so there is no logger to use
-          stderr.write(`Error: could not instantiate server from ${configFile}\n`);
-          stderr.write(`${error.stack}\n`);
+    const config = resolveAssetPath(params.config);
+    this.createComponentsManager(loaderProperties, config)
+      .then((componentsManager): Promise<void> =>
+        this.resolveVariables(componentsManager, argv).then(
+          (vars): Promise<void> => this.startApp(componentsManager, vars),
+          (error): void => this.writeError(stderr, `Could not load config variables from ${config}`, error),
+        ).catch((error): void => {
+          this.logger.error(`Could not start server: ${error}`);
           process.exit(1);
-        },
-      ).catch((error): void => {
-        this.logger.error(`Could not start server: ${error}`, { error });
-        process.exit(1);
-      });
+        }))
+      .catch((error): void => this.writeError(stderr, `Could not build the config files from ${config}`, error));
   }
 
   /**
-   * Creates the main app object to start the server from a given config.
-   * @param loaderProperties - Components.js loader properties.
-   * @param configFile - Path to a Components.js config file.
-   * @param variables - Variables to pass into the config file.
+   * Writes the given message and error to the `stderr` and exits the process.
    */
-  public async createApp(
-    loaderProperties: IComponentsManagerBuilderOptions<App>,
-    configFile: string,
-    variables: CliParams | Record<string, any>,
-  ): Promise<App> {
-    // Translate command-line parameters if needed
-    if (typeof variables.loggingLevel === 'string') {
-      variables = this.createVariables(variables as CliParams);
-    }
+  private writeError(stderr: WriteStream, message: string, error: Error): never {
+    stderr.write(`${message}\nCause:\n${createErrorMessage(error)}\n`);
+    stderr.write(`${error.stack}\n`);
+    process.exit(1);
+  }
 
-    // Set up Components.js
-    const componentsManager = await ComponentsManager.build(loaderProperties);
-    await componentsManager.configRegistry.register(configFile);
+  /**
+   * Handles the first Components.js instantiation,
+   * where CLI settings and variable mappings are created.
+   */
+  private async resolveVariables(componentsManager: ComponentsManager<CliResolver>, argv: string[]):
+  Promise<Record<string, unknown>> {
+    // Create a resolver, which combines a CliExtractor and a VariableResolver
+    const resolver = await componentsManager.instantiate(DEFAULT_CLI_RESOLVER, {});
+    // Extract values from CLI parameters
+    if (!resolver.extractor) {
+      throw new Error('No CliExtractor is defined.');
+    }
+    const cliValues = await resolver.extractor.handleSafe(argv);
+    //  Convert CLI values into variable mappings
+    return await resolver.resolver.handleSafe(cliValues);
```

Comment:
Not sure if Components.js is going to error on that, I will double check. Potentially it will only do this with the latest version that has stricter type checks.

Reply:
Right, Components.js is not going to throw here as the object used in the config is the `RecordObject` helper object that simply stores key/value combinations so that they can easily be extracted afterwards. To get correct typings we would have to create a new (tiny) class that takes an extractor/resolver as input arguments. Going to do that as it also creates a nicer config.